hail/src/main/scala/is/hail/expr/ir/analyses/SemanticHash.scala

```diff
@@ -0,0 +1,409 @@
+package is.hail.expr.ir.analyses
+
+import is.hail.backend.ExecuteContext
+import is.hail.expr.ir.functions.{TableCalculateNewPartitions, WrappedMatrixToValueFunction}
+import is.hail.expr.ir.{MatrixRangeReader, _}
+import is.hail.io.fs.FS
+import is.hail.io.vcf.MatrixVCFReader
+import is.hail.methods._
+import is.hail.types.virtual._
+import is.hail.utils.{Logging, TreeTraversal, toRichBoolean}
+import org.apache.commons.codec.digest.MurmurHash3
+
+import java.nio.ByteBuffer
+import scala.language.implicitConversions
+
+case object SemanticHash extends Logging {
+  type Type = Int
+  type NullableType = Integer
+
+  // Picked from https://softwareengineering.stackexchange.com/a/145633
+  def extend(x: Type, bytes: Array[Byte]): Type =
+    MurmurHash3.hash32x86(bytes, 0, bytes.length, x)
+
+
+  def apply(ctx: ExecuteContext)(root: BaseIR): Option[Type] =
+    ctx.timer.time("SemanticHash") {
+
+      // Running the algorithm on the name-normalised IR
+      // removes sensitivity to compiler-generated names
+      val nameNormalizedIR = ctx.timer.time("NormalizeNames") {
+        new NormalizeNames(_.toString, allowFreeVariables = true)(root)
+      }
+
+      val semhash = ctx.timer.time("Hash") {
+        def go: Option[Int] = {
+          var hash = MurmurHash3.DEFAULT_SEED
+
+          // Include an encoding of a node's trace to differentiate between
+          // IR trees that look identical when flattened
+          for ((ir, trace) <- levelOrder(nameNormalizedIR)) {
+            encode(ctx.fs, ir, trace) match {
+              case Some(bytes) => hash = extend(hash, bytes)
+              case None => return None
+            }
+          }
+
+          Some(hash)
+        }
+
+        go
+      }
+
+      log.info(s"IR Semantic Hash: $semhash")
+      semhash
+    }
+
+  private def encode(fs: FS, ir: BaseIR, trace: Int): Option[Array[Byte]] = {
+    val buffer =
+      Array.newBuilder[Byte] ++=
+        Bytes.fromClass(ir.getClass) ++=
+        Bytes.fromInt(trace)
+
+    ir match {
+      case a: AggExplode =>
```

Comment:
I'm not sure about leaving bound names like `a.name` here out of the hash. *Maybe* it's safe because of the normalize names, but I'm not confident about that, and anyways that could easily break after some innocuous looking future change. I think it's safer to just always include them.

Reply:
Ah, I thought `name` was a bound local so it didn't matter (in the same way that the name bound in a let doesn't matter)

Reply:
The name doesn't matter, in the sense that if you change it, and also change all captured references, the program is the same (that's [alpha equivalence](https://en.wikipedia.org/wiki/Lambda_calculus#Alpha_equivalence)). But the hash isn't aware of the binding structure. So for example if you ignore names at binding sites, these two IRs would hash the same:
```
Let('x', ..., Let('y', ..., ... Ref('x')))
Let('y', ..., Let('x', ..., ... Ref('x')))
```
It's possible that the NormalizeNames invariant prevents examples like these, by forcing the outer name to be lexicographically less than the inner or something like that, but I wouldn't want to depend on that without thinking it through carefully, and documenting the properties of NormalizeNames that this depends on. Much simpler to just hash the names.

Reply:
Thanks - this is a really good example

Reply:
Ok, after playing with this, I can't get an example to fail. I think normalizenames takes care of this for us.

NormalizeNames transforms your example to: 
```
Let('1', ..., Let('2', ..., ... Ref('1')))
Let('1', ..., Let('2', ..., ... Ref('2')))
```

For `Ref`, I include the bytes of the name being referenced in the semantic hash. I think the only time we'd see a hash collision with let/ref is if normalizenames produces a collision.